### PR TITLE
Dependency checks and environment setup assistance

### PR DESCRIPTION
**Pull request by petervdonovan**
This branch adds checks for various dependencies, with corresponding messages and suggested install commands in the UI. The new behavior is described mostly declaratively in `check_dependencies.ts`.

This feature has been exceedingly difficult to test. Tests exist, but coverage is incomplete. This is not good, but I think it is acceptable ~~because when the extension sends text to the VS Code terminal, it does not append a terminating newline. This means the user can review the suggested install/update command before pressing Enter~~ if CI is supplemented with manual tests.

### CODE REVIEW

**Comment by petervdonovan** [comment]
There are some tricky UI questions to be resolved here.

Currently, the user is presented with a notification that will typically have a button labeled either "View download page", "Update", or "Install". If they click a button that says "Update" or "Install", an update/install command is sent to the VS Code integrated terminal, which pops up on the user's screen. The problem is that the update/install command might not be what the user wants -- for example, maybe the user wants to manage their packages using `brew`, but this command bypasses that to do the installation using a bash script. **Therefore,** I decided to make the user press the <kbd>Enter</kbd> key to execute the command in the integrated terminal. Alternatives include:
* Providing more detail in the dialog/button so that the user knows in advance what command it will execute, then executing it in the terminal without waiting for the user to press <kbd>Enter</kbd>
* Showing a "[quick pick](https://github.com/microsoft/vscode-extension-samples/tree/main/quickinput-sample)" (sort of like a drop-down menu) that lists all the possible install commands, and lets the user choose one of them. This would have to happen in addition to showing a message explaining why the dependency is needed, because quick picks do not come with explanatory messages.

Another issue is what happens when the user performs an action that multiple dependencies depend on. It could be problematic for two dialogs to pop up at once, because the user may take several seconds to handle the first dialog, during which time the second dialog may disappear. (It takes 20 seconds for a dialog to disappear.)

**Comment by lhstrh** [comment]
> Currently, the user is presented with a notification that will typically have a button labeled either "View download page", "Update", or "Install". If they click a button that says "Update" or "Install", an update/install command is sent to the VS Code integrated terminal, which pops up on the user's screen. The problem is that the update/install command might not be what the user wants -- for example, maybe the user wants to manage their packages using `brew`, but this command bypasses that to do the installation using a bash script. **Therefore,** I decided to make the user press the Enter key to execute the command in the integrated terminal. Alternatives include:
> 
>     * Providing more detail in the dialog/button so that the user knows in advance what command it will execute, then executing it in the terminal without waiting for the user to press Enter
> 
>     * Showing a "[quick pick](https://github.com/microsoft/vscode-extension-samples/tree/main/quickinput-sample)" (sort of like a drop-down menu) that lists all the possible install commands, and lets the user choose one of them. This would have to happen in addition to showing a message explaining why the dependency is needed, because quick picks do not come with explanatory messages.
> 

I think that either of these alternatives are better than requiring the user to press Enter. I would go for the option that is easiest to implement. The key is that a user must be 1) given sufficient information to do a manual install and 2) given the opportunity to back out of an automatic install.
 
> Another issue is what happens when the user performs an action that multiple dependencies depend on. It could be problematic for two messages to pop up at once, because the user may take several seconds to handle the first message, during which time the second message may disappear. (It takes 20 seconds for a message to disappear.)

I'm not sure what to do about this one. Perhaps we shouldn't worry about it considering that a "retry" can always be performed to get to messages that disappeared?

**Comment by petervdonovan** [comment]
I still need to check in the GUI that this behaves as intended, but the tests pass, and I don't anticipate major changes in this PR before it gets merged.

**Comment by petervdonovan** [comment]
Another caveat: This PR introduces nondeterministic test failures like this:
```
Downloading VS Code 1.68.1 from https://update.code.visualstudio.com/1.68.1/darwin/stable
node:events:505
      throw er; // Unhandled 'error' event
      ^
Error: read ECONNRESET
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20)
Emitted 'error' event on ClientRequest instance at:
    at TLSSocket.socketErrorListener (node:_http_client:454:9)
    at TLSSocket.emit (node:events:527:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -54,
  code: 'ECONNRESET',
  syscall: 'read'
}
```

These only occur on the MacOS runners, but they occur pretty often -- something like half the time. I think that these failures are bad, but acceptable.

**Comment by petervdonovan** [comment]
Another caveat: This PR introduces nondeterministic test failures like this:
```
Downloading VS Code 1.68.1 from https://update.code.visualstudio.com/1.68.1/darwin/stable
node:events:505
      throw er; // Unhandled 'error' event
      ^
Error: read ECONNRESET
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20)
Emitted 'error' event on ClientRequest instance at:
    at TLSSocket.socketErrorListener (node:_http_client:454:9)
    at TLSSocket.emit (node:events:527:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -54,
  code: 'ECONNRESET',
  syscall: 'read'
}
```

These only occur on the MacOS runners, but they occur pretty often -- something like half the time. I think that these failures are bad, but acceptable.

**Comment by lhstrh** [comment]
Would caching be a good strategy to mitigate the connection issues?

File: src/version_checker.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+export const corepackVersionChecker = basicChecker(new Version('0.0.0'), 'corepack -v', false);
+export const curlVersionChecker = basicChecker(new Version('0.0.0'), 'curl -V', false);
+export const rustVersionChecker = basicChecker(config.rustVersion, 'rustc --version', false);
+export const cmakeVersionChecker = basicChecker(config.rustVersion, 'cmake --version', false);
+export const nvmVersionChecker = basicChecker(new Version('0.0.0'), 'nvm --version', false);
+export const snapVersionChecker = basicChecker(new Version('0.0.0'), 'snap --version', false);
+export const aptGetVersionChecker = basicChecker(new Version('0.0.0'), 'apt-get -v', false);
+export const chocolateyVersionChecker = basicChecker(new Version('0.0.0'), 'choco -v', false);
+export const rtiVersionChecker = async () => {
+    // TODO: Update when #1233 is addressed: https://github.com/lf-lang/lingua-franca/issues/1233
```
Comment (lhstrh) [inline src/version_checker.ts:86]: https://github.com/lf-lang/lingua-franca/issues/1233 is closed...

Comment (petervdonovan) [reply src/version_checker.ts:86]:
```suggestion
    // TODO: Update when #76 is addressed: https://github.com/lf-lang/reactor-c/issues/76
```

File: src/build_lds.ts [hunk 2]
```diff
@@ -1,22 +1,20 @@
...
+
 'use strict';
 
 import * as fs from 'fs'
 import * as path from 'path'
 import * as rimraf from 'rimraf'
 import simpleGit, { SimpleGit } from 'simple-git'
 import { Command, OptionValues } from 'commander'
-import { Config } from './config'
+import * as config from './config'
```
Comment (lhstrh) [inline src/build_lds.ts:13]: Why is this better?

**Review by lhstrh — APPROVED** [review]
I have no tested this locally yet, but this looks great! I left some minor comments.

File: src/build_lds.ts (hunk 2 above)
Comment (lhstrh) [reply src/build_lds.ts:13]: You might have a similar reaction to the reactor-ts code base. I would enjoy receiving some code review from you on that :-)

**Comment by petervdonovan** [comment]
After the latest bugfix, I just got this branch to pass a basic smoke test (by installing Rust from within the VS Code editor). I think that what we have now should be better than nothing.

File: src/build_commands.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import * as vscode from 'vscode';
 import { LanguageClient } from 'vscode-languageclient';
+import { getTerminal, MessageShower } from './utils';
```
[the remark this answers is not in the document]
Comment (petervdonovan) [reply src/build_commands.ts:3]: Well, it is a function that shows messages, and noun-ifying it as a "Display" makes it sound like something more concrete, like a box that appears on the screen. Maybe the least confusing choice would be to keep it as a verb.

[the remark this answers is not in the document]
Comment (lhstrh) [reply src/build_commands.ts:3]: I'm fine with a verb, it's just that shower (:shower:) is also a noun.

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
+{
+    "version": "0.2.0",
```
[the remark this answers is not in the document]
Comment (petervdonovan) [reply .vscode/launch.json:2]: Both this version number and the version number in `tasks.json` were auto-generated by VS Code. They specify the version of the format of the JSON file.

This file is used for debugging. It lets you start an instance of VS Code from the debugging tab on the left while debugging it with breakpoints. For some reason, it needs `tasks.json` to do this.

File: src/build_commands.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lhstrh) [reply src/build_commands.ts:3]: How about `MessageDisplayHelper`?

[the remark this answers is not in the document]
Comment (petervdonovan) [reply src/build_commands.ts:3]: That sounds good to me!

**Pull request merged**